src/memray/reporters/assets/temporal_flamegraph.js

```diff
@@ -0,0 +1,299 @@
+import { debounced } from "./common";
+
+import {
+  initThreadsDropdown,
+  drawChart,
+  handleFragments,
+  onFilterUninteresting,
+  onFilterImportSystem,
+  onFilterThread,
+  onResetZoom,
+  onResize,
+  onInvert,
+  getFilteredChart,
+  getFlamegraph,
+} from "./flamegraph_common";
+
+var active_plot = null;
+var current_dimensions = null;
+
+var parent_index_by_child_index = (function () {
+  let ret = new Array(packed_data.nodes.children.length);
+  console.log("finding parent index for each node");
+  for (const [parentIndex, children] of packed_data.nodes.children.entries()) {
+    children.forEach((idx) => (ret[idx] = parentIndex));
+  }
+  console.assert(ret[0] === undefined, "root node has a parent");
+  return ret;
+})();
+
+function packedDataToTree(packedData, rangeStart, rangeEnd) {
+  const { strings, nodes, unique_threads } = packedData;
+
+  console.log("constructing nodes");
+  const node_objects = nodes.name.map((_, i) => ({
+    name: strings[nodes["name"][i]],
+    location: [
+      strings[nodes["function"][i]],
+      strings[nodes["filename"][i]],
+      nodes["lineno"][i],
+    ],
+    value: 0,
+    children: nodes["children"][i],
+    n_allocations: 0,
+    thread_id: strings[nodes["thread_id"][i]],
+    interesting: nodes["interesting"][i] !== 0,
+    import_system: nodes["import_system"][i] !== 0,
+  }));
+
+  console.log("mapping child indices to child nodes");
+  for (const [parentIndex, node] of node_objects.entries()) {
+    node["children"] = node["children"].map((idx) => node_objects[idx]);
+  }
+
+  // We could binary search rather than using a linear scan...
+  console.log("finding leaked allocations");
+  packedData.intervals.forEach((interval) => {
+    let [allocBefore, deallocBefore, nodeIndex, count, bytes] = interval;
+
+    if (
+      allocBefore >= rangeStart &&
+      allocBefore <= rangeEnd &&
+      (deallocBefore === null || deallocBefore > rangeEnd)
+    ) {
+      while (nodeIndex !== undefined) {
+        node_objects[nodeIndex].n_allocations += count;
+        node_objects[nodeIndex].value += bytes;
+        nodeIndex = parent_index_by_child_index[nodeIndex];
+      }
+    }
+  });
```

Comment:
A possible implementation of this using binary search (not tested):

```
console.log("finding leaked allocations");

packedData.intervals.forEach((interval) => {
  let [allocBefore, deallocBefore, nodeIndex, count, bytes] = interval;

  if (
    allocBefore >= rangeStart &&
    allocBefore <= rangeEnd &&
    (deallocBefore === null || deallocBefore > rangeEnd)
  ) {
    // Use binary search to find the parent index of the node
    let low = 0;
    let high = parent_index_by_child_index.length - 1;
    while (low <= high) {
      let mid = Math.floor((low + high) / 2);
      if (nodeIndex > parent_index_by_child_index[mid]) {
        low = mid + 1;
      } else if (nodeIndex < parent_index_by_child_index[mid]) {
        high = mid - 1;
      } else {
        nodeIndex = mid;
        break;
      }
    }

    // Update the node_objects for each parent node
    while (nodeIndex !== undefined) {
      node_objects[nodeIndex].n_allocations += count;
      node_objects[nodeIndex].value += bytes;
      nodeIndex = parent_index_by_child_index[nodeIndex];
    }
  }
});
```

Reply:
To be honest it looks more complex than the linear search and I don't think it gives us that much 🤷

Reply:
Hm. We could pull in d3-array and use https://github.com/d3/d3-array/blob/main/README.md#bisectLeft perhaps.

The linear search is fast enough using the samples I've tried with, but I'm not sure if that would be the case with a longer run or a huge capture file.